<a href="https://colab.research.google.com/github/MartinCano2004/Analitica-de-Negocios/blob/main/Reto2_%C3%81rboles_de_Decisi%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**RETO 2: ÁRBOLES DE DECISIÓN**

Martín Cano Aristizábal

**DESCRIPCIÓN DE VARIABLES**

- Business: Tipo de negocio (si es Software (1) o Hardware (0)).
- Age: Edad del empleado.
- Female: Si el empleado es mujer (1) o es hombre (0)
- Years: Los años que el empleado lleva trabajando.
- Certificates: Si el empleado tiene certificados y cuántos son.
- Feedback: El puntaje de retroalimentación promedio que cada empleado recibede sus compañeros y supervisor en la evaluación anual de 360 grados. Las puntuaciones posibles van de 0 a 4.
- Salary: El salario del empleado.
- NPS: El puntaje neto del promotor es un indicador clave de la satisfacción y lealtad del cliente
- College: El empleado posee título universitario. SI(1) NO(0)

**0. Cargar las librerías de trabajo.**

In [ ]:
import numpy as np
import pandas as pd

#para implementación y evalución del modelo:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix


**1. Cargar la base de datos.**

In [ ]:
nxl="/content/2. TechSalesReport_Int_L.xlsx"
XDB=pd.read_excel(nxl,sheet_name=0)
XDB['College'] = XDB['College'].replace({'Yes': 1, 'No': 0})
print(XDB)

#Definir XD - yd
XD=XDB[['Business','Age','Years','Female','Feedback','Salary','Certificates','NPS']]
yd=XDB['College']

       Business  Age  Female  Years  College  Certificates  Feedback  Salary  \
0             0   59       1      2        1             1      2.01   70200   
1             0   52       0     10        1             4      3.64  133000   
2             1   47       1      1        1             1      3.88   52600   
3             0   61       0      2        1             3      2.70   96000   
4             1   39       0      1        0             5      3.44  122000   
...         ...  ...     ...    ...      ...           ...       ...     ...   
21985         1   44       0      1        1             4      1.80  128000   
21986         1   23       1      6        1             6      1.77   53400   
21987         0   48       1      4        1             0      2.46   50400   
21988         1   29       0      4        1             2      3.68   70000   
21989         1   23       1      2        1             1      2.13   47400   

       NPS  
0        5  
1       10  


<ipython-input-4-dacafb31eb1b>:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  XDB['College'] = XDB['College'].replace({'Yes': 1, 'No': 0})


**2. Implementamos el modelo.**

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)

#Se procede con el despliegue del árbol de decisión.
from sklearn.tree import export_graphviz #Exportar.
from pydotplus import graph_from_dot_data #Crear el gráfico.

vs=['Business','Age','Years','Female','Feedback','Salary','Certificates','NPS']

dot_data=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_data)
graph.write_png('Reto2_Arbol.png')

True

**3. Desempeño del Modelo.**

In [ ]:
ydp=mar.predict(XD)
cm=confusion_matrix(yd,ydp)
print(cm)


VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

Exactitud=((VP+VN)/(VP+VN+FP+FN)) #Funcionamiento general del modelo
print('La Exactitud del modelo es del',Exactitud)
Sensibilidad=(VP/(VP+FN)) #Como se comporto el modelo con los positivos
print('La Sensibilidad del modelo es del',Sensibilidad)
Especificidad=(VN/(VN+FP)) #Como se comporto el modelo con los 0
print('La Especificidad del modelo es del',Especificidad)
Precision=(VP/(VP+FP)) #Cual es el acierto clasificando los positivos
print('La Precision del modelo es del',Precision)
TasaError=((FP+FN)/(VP+VN+FP+FN)) #Equivocación general
print('La Tasa de Error del modelo es del',TasaError)
PredNeg=(VN/(VN+FN)) #Como funciona el modelo calificando en 0.
print('La PredNeg es del',PredNeg)

[[  114  4356]
 [   69 17451]]
La Exactitud del modelo es del 0.7987721691678036
La Sensibilidad del modelo es del 0.9960616438356165
La Especificidad del modelo es del 0.025503355704697986
La Precision del modelo es del 0.8002476269087908
La Tasa de Error del modelo es del 0.20122783083219645
La PredNeg es del 0.6229508196721312


**4. Implementar Naive Bayes.**

In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB()
mnb.fit(XD,yd.values.ravel())

GaussianNB()

In [ ]:
from time import ctime
#Se obtiene la información del modelo
ct=mnb.class_count_
print("¿Como estan separados los datos?\n",ct)
print("El total de los datos es:\n",len(XD))

¿Como estan separados los datos?
 [ 4470. 17520.]
El total de los datos es:
 21990


**5. Llevar a cabo el pronóstico del título Universitario para las personas agrupadas en la base de datos de pronóstico.**

In [ ]:
XDB2=pd.read_excel(nxl,sheet_name=1)
XDB2.head(100) #Solo muestra 100 datos

# Convertir Business a valores numericos
XDB2['Business']=pd.Categorical(XDB2['Business']).codes
XDB2=XDB2.dropna()
print(XDB2)

dfp=mnb.predict

    Sales_Rep  Business  Age  Female  Years Personality  Certificates  \
0           1         0   63       0      2    Sentinel             5   
1           2         1   38       1      2    Diplomat             4   
2           3         1   47       0      1     Analyst             2   
3           4         1   25       0      1    Explorer             2   
4           5         0   49       0     10    Explorer             4   
5           6         1   48       0      1    Diplomat             1   
6           7         1   25       1      1    Explorer             6   
7           8         0   39       0      1    Diplomat             2   
8           9         1   29       0      1    Explorer             5   
9          10         1   45       1      3    Diplomat             3   
10         11         0   35       0      6    Diplomat             1   
11         12         0   30       0      4    Diplomat             2   
12         13         1   28       0      1     Ana

**6. Número de personas y de nodos puros.**

In [ ]:
# Total de personas:
total_personas = len(XDB)

# Número de personas con título universitario (inscritas):
num_inscritas = XDB['College'].sum()

# Número de personas sin título universitario (no inscritas):
num_no_inscritas = total_personas - num_inscritas

# Imprimir los resultados:
print("Total de personas en la base de datos:",total_personas)
print("Número de personas con título universitario (inscritas):",num_inscritas)
print("Número de personas sin título universitario (no inscritas):",num_no_inscritas)

# Determinar el número de nodos puros:
# Obtener el número total de nodos en el árbol:
n_nodes = mar.tree_.node_count

# Obtener la información de las salidas por izquierda y derecha de cada nodo:
children_left = mar.tree_.children_left
children_right = mar.tree_.children_right

# Contar los nodos puros (nodos hoja donde todos los registros pertenecen a la misma clase):
pure_nodes = sum(1 for i in range(n_nodes) if children_left[i] == children_right[i])

# Imprimir el resultado:
print("Número de nodos puros:",pure_nodes)

Total de personas en la base de datos: 21990
Número de personas con título universitario (inscritas): 17520
Número de personas sin título universitario (no inscritas): 4470
Número de nodos puros: 16


**ANÁLISIS DE RESULTADOS.**

Para la implementación del modelo de Árbol de Decisión se tomaron un total de 9 variables, 8 variables de entrada entre las que se destacan Business, Age, Years, Female, Feedback, Salary, Certificates y NPS; y una variable de salida que permite evaluar si el empleado tiene o no un título universitario. Para la construcción del modelo se tuvieron un total de 21990 datos divididos en 17520 personas contítulo universitario y un total de 4470 sin título. Luego de la ejecución del modelo se logró una **exactitud cercana al 80%**, lo que está por encima del 75% frente a la exactitud, indicando la calidad del modelo en el pronóstico. Es importante destacar que el **coeficiente de sensibilidad fue cercano al 100%**, mientras que **la Precisión del Modelo fue del 80%**, lo que corrobora el buen funcionamiento del mismo.